### DISCUSSION TITLE
Mobile-friendly UI design

### DISCUSSION THREAD

**Discussion by poyribn** [godotengine/godot-proposals#4326] [Editor]
> The current version doesn't have a lot of mobile specific changes, so it's only really usable on a tablet with keyboard and mouse - but the foundation is there to improve upon, and we're interested in your feedback and ideas son how to make the Godot experience more mobile friendly!
> 	 	 — [ Godot Website -> News:
Dev snapshot: Godot 3.5 beta 3 § Android editor port (GH-57747)
](https://godotengine.org/article/dev-snapshot-godot-3-5-beta-3)

This thread is for discussing how we could make the Godot Editor more mobile-friendly, if you have ideas.

I already have 2 ideas:

- A very rough draft of a good-for-narrow-screens design of the project manager in the form of [a Godot Project](https://github.com/godotengine/godot-proposals/files/8398885/mobile-friendly-project-manager-draft.zip) (this design is bad for medium and wide screens, so if actually putting it into Godot, Godot should switch between this narrow design and the current wide design depending on screen size).
The Editor itself is very unlikely to become mobile-friendly except on Tablets, so we don't have to worry about this anytime soon.
- Native scaling in Godot. As more and more of the screens used are high-density (even on Desktop), it gets more and more important to support this (especially if you want to support Touchscreen Devices, where if the buttons are physically too small it gets very hard to press them as you can't precisely move a cursor like you can with the mouse). Godot already has scaling, but to make it more mobile-friendly a kind of realtime scaling might be needed that works basically like the [DPIScaleControl](https://codeberg.org/sosasees/godot-4-responsive-ui-plugin#dpiscalecontrol) in my [responsive UI plugin](https://codeberg.org/sosasees/godot-4-responsive-ui-plugin). Putting this into the Engine might also encourage more developers to use DPI&#x2011;aware scaling because that would no longer need inserting a DPI&nbsp;scaling node.

**Comment by Zireael07** [comment]
+1 to having a narrow design for the editor.

**Reply by poyribn** [reply]
Yes: To me, this seems useful even on my 1920×1080px monitor.
It's challenging to design, but if done right it'd be very useful.

**Reply by Zireael07** [reply]
Same here. If I'm looking up stuff in the browser while coding, which happens quite frequently, the Godot window is narrower and I often have very little space left...

**Reply by lentsius-bark** [reply]
I wonder how this could be done. A heavier use of tabs? Allowing for each panel to be placed at the bottom (in other words any tab could be anywhere)?
Thinking about it now, my colleagues have used unity on portrait monitors and what allowed then to do so was how you can out any window anywhere

**Reply by Zireael07** [reply]
In my experience the two things that would make life better for a narrow screen are:
a) move the script list from vertical to horizontal (think VS Code) - there is already a proposal for that
b) allow hiding/folding both the scene tree (to the left) and the inspector (to the right) like many mobile/touchscreen apps do (drag to the edge to hide, drag to center to unhide, and possibly hotkeys on desktop)

**Reply by Mickeon** [reply]
See also my related Discussion & Proposal:
https://github.com/godotengine/godot-proposals/discussions/4248

**Comment by lentsius-bark** [comment]
I really like the scaling idea, I'm imagining a slider in the viewport top panel that would allow the user to specify the scaling they want to be seeing/using

**Comment by Calinou** [comment]
> Native scaling in Godot.

[This is already supported](https://docs.godotengine.org/en/stable/tutorials/viewports/multiple_resolutions.html) :slightly_smiling_face: 

Godot's stretch modes are pretty good as they are. I don't see much that can be done to improve it, other than adding options for integer scaling (which are not relevant for the editor itself). If you need more control over scaling, you can use the [2D scale factor property](https://github.com/godotengine/godot/pull/52170) which was also [cherry-picked to 3.4](https://github.com/godotengine/godot/pull/52137). It can be used on top of the `canvas_items` stretch mode (formerly `2d`), or with the `viewport` stretch mode, or with the `disabled` stretch mode.

However, the editor scale does **not** rely on the `canvas_items` stretch mode. Instead, it uses a different editor theme generated for the set scale, which means that the scale can't be changed without restarting the editor. On the bright side, this allows icons to be generated for the display's native resolution when the editor starts. If we relied on scaling, icons would end up looking pixelated or blurry at high scale factors (or at low scale factors, if you chose to always use a high scale factor when generating icons).

One way to improve touchscreen usability is to allow buttons to have a touch input margin. This would allow buttons to be triggered by touch input even if it's done slightly outside the button's visual appearance. For small buttons, this greatly improves usability. This can already be worked around by making buttons larger in the first place, but not requiring you to change button sizes would make this change easier to adopt in existing desktop → mobile game ports (including the Godot editor).

**Edit:** I opened a proposal for touch input margin: https://github.com/godotengine/godot-proposals/issues/4338

**Comment by Mesum-Hussain** [comment]
My Ideas to improve the UI of Godot Android app are:
 
1) Make Visual Scripting the central feature of the app and design the app UI around it, such that Visual Scripting becomes very easy.
Game developers should rarely need to use the keyboard on Tablets and phones.

2) The app should be given a design overhaul to look like a native Android app instead of looking like a desktop app ported to Android.

**Reply by Mickeon** [reply]
![image](https://user-images.githubusercontent.com/66727710/191547212-2582bb12-f371-405d-9805-40e08dea32dd.png)

**Reply by Calinou** [reply]
For context: https://godotengine.org/article/godot-4-will-discontinue-visual-scripting

> 2. The app should be given a design overhaul to look like a native Android app instead of looking like a desktop app ported to Android.

This is not possible as it requires too much maintenance (especially if this implies that the app should use native Android controls instead of mimicking them).

**Reply by Mesum-Hussain** [reply]
> For context: https://godotengine.org/article/godot-4-will-discontinue-visual-scripting
> 
> > 2. The app should be given a design overhaul to look like a native Android app instead of looking like a desktop app ported to Android.
> 
> This is not possible as it requires too much maintenance (especially if this implies that the app should use native Android controls instead of mimicking them).

I am not aware of the complexity of this task, but what I was referring to was to make the UI more touchscreen friendly, by making the touch targets bigger and introducing multi-touch swipe gestures.

I am sure making the Godot engine more touchscreen friendly will definitely help with various devices including iPads (if ported), Android phones (especially Foldable Phones) and Tablets, Windows touchscreen devices and all the Chromebooks out there. We can even make it such that those gestures work on a laptop's trackpad too. 

And I would also prefer if Godot UI became more modern and good looking like Android apps rather than looking like an old desktop app.

**Reply by Mesum-Hussain** [reply]
> ![image](https://user-images.githubusercontent.com/66727710/191547212-2582bb12-f371-405d-9805-40e08dea32dd.png)
> 

I am aware, but I think this is a good reason not to.

They can make VisualScript better and more integrated into Godot and design a good touchscreen UI surrounding it so that the user rarely needs to use the keyboard to make their game.

Porting this to iPad and Android will increase the popularity of both VisualScript and Godot drastically..... because as of now, people are desperate to use the M1 chip in their iPads to its full potential but don't have apps.
Godot being the only game engine available on iPad will make it very popular amongst iPad owners. The same thing applies to Android Tablets and Foldable phones too.

**Reply by poyribn** [reply]
i agree that making Godot more touchscreen-friendly would be nice but shouldn't be a top priority.
most phone/tablet ports are by necessity dumbed-down compared to their desktop versions because touch input is more limited than keyboard and mouse, so making a separate touch edition of Godot would increase maintainance cost at little-to-no benefit.

also, i feel like i would still find GDScript faster to type with my phone keyboard rather than VisualScript, because today's phones are big enough for a good touch keyboard (smaller phones are not made anymore because the 4Ah batteries that we demand don't fit inside them).

at its core, Godot is a desktop program that should not be compromised to work better on touch devices. i think most people are fine with having to attach a keyboard and mouse if it means that they can use an uncompromised game engine, and would even prefer that to a compromised touch-adjusted version.

**Reply by Mesum-Hussain** [reply]
> most phone/tablet ports are by necessity dumbed-down compared to their desktop versions because touch input is more limited than keyboard and mouse, 

I disagree on this. I think that the reason for the mobile apps being dumbed-down wasn't because that the touch UI is not as capable as keyboard+mouse, rather because we didn't had powerful enough chips and phones generally had an image of being a low power portable computer with small screen realestate compared to a desktop workstation so people weren't actually expecting full fledged features as well.
But now, Tablets like iPad comes with the same chip as a MacBook and so many Pro apps which were previously only available on desktop are available for Tablets so the perception of people regarding the capability of their portable devices is also changing, and it has a considerably big screen to work with too. 
Now that my Tablet is running on a very powerful chip and has 4 times more RAM than my 5 years old laptop, I really wonder if my Tablet should be provided with a dumbed-down version of the app my 5 years old laptop is running. 
I also wonder why you think Touch + Virtual Keyboard interface is not as powerful as a Mouse + Physical Keyboard interface when both do pretty much the same thing, and the Touch + Virtual Keyboard provides you with the gestures too. I think it's the design failure that we are not able to utilise the Touch + Virtual Keyboard interface to its full potential, not any inherent incapability of the interface itself.

> also, i feel like i would still find GDScript faster to type with my phone keyboard rather than VisualScript

And my suggestion was to try to eliminate the typing part entirely or reduce it as much as possible. I am not trying to simply promote VisualScript for no reason, I just think it would help in reducing the need of a keyboard input significantly compared to GDScript so that's why I am suggesting it. If we can come up with an alternative to VisualScript which helps achieving this same goal, I am fine with dropping VisualScript in favour of it. All I am suggesting is that Tablet app should be made Touch first, rather than keyboard first..... Touch input should be dominant in the workflow.

> at its core, Godot is a desktop program that should not be compromised to work better on touch devices.

And I don't see any reason why any Godot features would get compromised. We just need to design the interface in such a way that it is touchscreen friendly and at the same time offers each and every single feature that the current Godot app offers.

>i think most people are fine with having to attach a keyboard and mouse if it means that they can use an uncompromised game engine, and would even prefer that to a compromised touch-adjusted version.

And I think, if we make an uncompromised app for the Tablets, even the traditional desktop app users would prefer using Godot on their Tablet if they have one. It would simply be more portable and simple to use and understand, than a desktop app running in a laptop. 
That makes me feel like it's totally worth it to give Godot a design overhaul to make it mobile-first app.
Tablets and Foldable phones are going to be a big market a couple of years down the line, we wouldn't go wrong taking them seriously right now.

**Reply by poyribn** [reply]
@Mesum-Hussain Thanks, you convinced me that an uncompromised tablet version of Godot is possible and makes sense.

i think the first step towards this should be to change the scripts list in the "Script" tab from vertical tabs to horizontal tabs to make Godot friendlier to small screens in general:
https://github.com/godotengine/godot-proposals/discussions/4248

**Reply by Mickeon** [reply]
Do note that the proposal should ideally be an optional setting. [The mere suggestion of it is controversial (see reception)](https://github.com/godotengine/godot-proposals/issues/4081). It would also imply that both "vertical" and horizontal tabs would have to be supported, at the same time, which may prove tricky.

**Comment by DigitalMan171** [comment]
I know this is a thread from a while back, but is it still active, or is there something current? I have some simple suggestions that may be great for improving mobile useability.

**Reply by Zireael07** [reply]
It's still open, so add your suggestions here

**Reply by poyribn** [reply]
yes, i still welcome new comments in this discussion.

in its starting comment,
i just noticed links which should lead to my "system scaling in Godot" repository but i have deleted it since,
so i edited the links to lead to my "responsive UI plugin" repository which has a similar DPI&#x2011;scaling node

**Reply by DigitalMan171** [reply]
Okay, thanks, and sorry, I created a suggestion thread covering some of the ideas. My main consideration involves use of tabs, possibly on the sidebar, and occasionally portrait mode for typing and scrolling docks. I'm guessing testing modifications to the UI would involve compiling the program itself?

**Reply by poyribn** [reply]
it is okay. i am sure a moderator can merge these discussions.

&#x203b; other discussion: https://github.com/godotengine/godot-proposals/discussions/13327
